File: packages/modules/npm.py
import logging
import requests
import semver
import aiohttp
from requests.models import Response

from packages.models import VersionedPackage

NPM_REGISTRY_URL = "https://registry.npmjs.org"


async def get_package(name: str, range: str) -> VersionedPackage:
    url = f"{NPM_REGISTRY_URL}/{name}"
    async with aiohttp.ClientSession() as session:
        try:
            async with session.get(url) as response:
                response.raise_for_status()
                npm_package = await response.json()
                versions = list(npm_package["versions"].keys())
                version = semver.min_satisfying(versions, range)
                if version not in npm_package["versions"]:
                    raise aiohttp.ClientResponseError(
                        status=404, 
                        message="Version not found", 
                        request_info=response.request_info,
                        history=response.history
                    )
                version_record = npm_package["versions"].get(version)
                package = VersionedPackage(
                    name=version_record["name"],
                    version=version_record["version"],
                    description=version_record["description"],
                )
                dependencies = version_record.get("dependencies", {})

                package.dependencies = [
                    await get_package(name=dep_name, range=dep_range)
                    for dep_name, dep_range in dependencies.items()
                ]
                
        except aiohttp.ClientResponseError as http_err:
            logging.error(f"HTTP error occurred: {http_err}")
            raise
        except Exception as err:
            logging.error(f"Other error occurred: {err}")
            raise

    return package


def request_package(name: str, range: str) -> tuple[VersionedPackage, dict]:
    url = f"{NPM_REGISTRY_URL}/{name}"

    npm_package = requests.get(url).json()

    versions = list(npm_package["versions"].keys())
    version = semver.min_satisfying(versions, range)
    version_record = npm_package["versions"][version]

    return VersionedPackage(
        name=version_record["name"],
        version=version_record["version"],
        description=version_record["description"],
    ), version_record.get("dependencies", {})

File: src/npm.cc
#include "model.h"
#include "util.h"

#include <string>
#include <sstream>
#include <nlohmann/json.hpp>

namespace npm {
  namespace {
    using model::VersionedPackage;
    constexpr std::string_view NPM_REGISTRY_URL = "https://registry.npmjs.org";
  }

  VersionedPackage get_versioned_package(const std::string& name, const std::string& range) {
    std::string url = std::string(NPM_REGISTRY_URL) + "/" + name;

    std::ostringstream response_stream;
    nlohmann::json npm_package = nlohmann::json::parse(util::curl_request(url, &response_stream));
    response_stream.clear();

    std::vector<std::string> versions;
    for (auto it = npm_package["versions"].begin(); it != npm_package["versions"].end(); ++it) {
        versions.push_back(it.key());
    }
    std::string version = util::max_satisfying(versions, range);
    const nlohmann::json& version_record = npm_package["versions"][version];
    const nlohmann::json& dependencies = version_record.value("dependencies", nlohmann::json::object());

    VersionedPackage package{
        version_record["name"],
        version_record["version"],
        version_record["description"],
        {}
    };

    for (const auto& dep : dependencies.items()) {
        package.add_dependency(get_versioned_package(dep.key(), dep.value()));
    }

    return package;
  }
}